This patch is based on https://github.com/libssh2/libssh2/pull/1641

It can be removed when we update to next libssh2 release

--- libssh2-1.11.1/src/knownhost.c
+++ libssh2-1.11.1/src/knownhost.c
@@ -142,8 +142,8 @@ knownhost_add(LIBSSH2_KNOWNHOSTS *hosts,
     struct known_host *entry;
     size_t hostlen = strlen(host);
     int rc;
-    char *ptr;
-    size_t ptrlen;
+    char *ptr = NULL;
+    size_t ptrlen = 0;
 
     /* make sure we have a key type set */
     if(!(typemask & LIBSSH2_KNOWNHOST_KEY_MASK))
@@ -175,13 +175,30 @@ knownhost_add(LIBSSH2_KNOWNHOSTS *hosts,
                                     host, hostlen);
         if(rc)
             goto error;
+
+        if(!ptr || ptrlen == 0) {
+            rc = _libssh2_error(hosts->session, LIBSSH2_ERROR_INVAL,
+                                "Base64 decoded value is invalid");
+            goto error;
+        }
+
+
         entry->name = ptr;
         entry->name_len = ptrlen;
 
+        ptr = NULL;
+        ptrlen = 0;
         rc = _libssh2_base64_decode(hosts->session, &ptr, &ptrlen,
                                     salt, strlen(salt));
         if(rc)
             goto error;
+
+        if(!ptr || ptrlen == 0) {
+            rc = _libssh2_error(hosts->session, LIBSSH2_ERROR_INVAL,
+                                "Base64 decoded value is invalid");
+            goto error;
+        }
+
         entry->salt = ptr;
         entry->salt_len = ptrlen;
         break;
--- libssh2-1.11.1/src/misc.c
+++ libssh2-1.11.1/src/misc.c
@@ -453,9 +453,6 @@ size_t _libssh2_base64_encode(LIBSSH2_SE
     *outptr = NULL; /* set to NULL in case of failure before we reach the
                        end */
 
-    if(insize == 0)
-        insize = strlen(indata);
-
     base64data = output = LIBSSH2_ALLOC(session, insize * 4 / 3 + 4);
     if(!output)
         return 0;
